File: src/test/java/net/stefanhahmann/polyhedron/StarConvexPolyhedronTest.java
package net.stefanhahmann.polyhedron;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class StarConvexPolyhedronTest
{
	StarConvexPolyhedron unitPolyhedronAtZero;

	StarConvexPolyhedron tinyPolyhedronAtZero;

	StarConvexPolyhedron bigPolyhedronAt50;

	private final double[] centerZero = new double[] { 0, 0, 0 };

	double[] center50 = new double[] { 50, 50, 50 };

	private double[] pointInsideTinyPolyhedron;

	private double[] pointOutSideTinyPolyhedron;

	double[] shouldBeOutside = new double[] { 25, 12, 50 };

	private final Random random = new Random();

	@BeforeEach
	public void setUp()
	{
		random.setSeed( 1 );

		List< Double > distances = new ArrayList<>( Collections.nCopies( 96, 1d ) );
		unitPolyhedronAtZero = new StarConvexPolyhedron( centerZero, distances );

		distances = random.doubles( 96, 2, 5 )
				.boxed()
				.collect( Collectors.toList() );
		tinyPolyhedronAtZero = new StarConvexPolyhedron( centerZero, distances );
		distances = random.doubles( 96, 5, 40 )
				.boxed()
				.collect( Collectors.toList() );
		bigPolyhedronAt50 = new StarConvexPolyhedron( center50, distances );
		pointInsideTinyPolyhedron = new double[] { 1, 1, 1 };
		pointOutSideTinyPolyhedron = new double[] { 10, 20, 15 };
	}

	@Test
	void testContains1()
	{
		assertTrue( tinyPolyhedronAtZero.contains( centerZero ) );
		assertTrue( tinyPolyhedronAtZero.contains( pointInsideTinyPolyhedron ) );
		assertFalse( tinyPolyhedronAtZero.contains( pointOutSideTinyPolyhedron ) );
	}

	@Test
	void testContains()
	{
		random.setSeed( 1 );
		assertTrue( bigPolyhedronAt50.contains( center50 ) );
		assertFalse( bigPolyhedronAt50.contains( shouldBeOutside ) ); //
	}

	@Test
	void testGetBoundingBox3D()
	{
		StarConvexPolyhedron.BoundingBox3D boundingBox3D = tinyPolyhedronAtZero.getBoundingBox3D();
		assertTrue( boundingBox3D.contains( centerZero ) );
		assertTrue( boundingBox3D.contains( pointInsideTinyPolyhedron ) );
		assertFalse( boundingBox3D.contains( pointOutSideTinyPolyhedron ) );
	}
}

File: src/test/kotlin/net/stefanhahmann/polyhedron/kotlin/StarConvexPolyhedronTest.kt
package net.stefanhahmann.polyhedron.kotlin

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.*
import java.util.stream.Collectors

internal class StarConvexPolyhedronTest {
    private lateinit var unitPolyhedronAtZero: StarConvexPolyhedron

    private lateinit var tinyPolyhedronAtZero: StarConvexPolyhedron

    private lateinit var bigPolyhedronAt50: StarConvexPolyhedron

    private val centerZero = doubleArrayOf(0.0, 0.0, 0.0)

    private var center50: DoubleArray = doubleArrayOf(50.0, 50.0, 50.0)

    private lateinit var pointInsideTinyPolyhedron: DoubleArray

    private lateinit var pointOutSideTinyPolyhedron: DoubleArray

    private var shouldBeOutside: DoubleArray = doubleArrayOf(25.0, 12.0, 50.0)

    private val random = Random()

    @BeforeEach
    fun setUp() {
        random.setSeed(1)

        var distances: List<Double>? = ArrayList(Collections.nCopies(96, 1.0))
        unitPolyhedronAtZero = StarConvexPolyhedron(centerZero, distances)

        distances = random.doubles(96, 2.0, 5.0)
            .boxed()
            .collect(Collectors.toList())
        tinyPolyhedronAtZero = StarConvexPolyhedron(centerZero, distances)
        distances = random.doubles(96, 5.0, 40.0)
            .boxed()
            .collect(Collectors.toList())
        bigPolyhedronAt50 = StarConvexPolyhedron(center50, distances)
        pointInsideTinyPolyhedron = doubleArrayOf(1.0, 1.0, 1.0)
        pointOutSideTinyPolyhedron = doubleArrayOf(10.0, 20.0, 15.0)
    }

    @Test
    fun testContains1() {
        Assertions.assertTrue(tinyPolyhedronAtZero.contains(centerZero))
        Assertions.assertTrue(tinyPolyhedronAtZero.contains(pointInsideTinyPolyhedron))
        Assertions.assertFalse(tinyPolyhedronAtZero.contains(pointOutSideTinyPolyhedron))
    }

    @Test
    fun testContains() {
        random.setSeed(1)
        Assertions.assertTrue(bigPolyhedronAt50.contains(center50))
        Assertions.assertFalse(bigPolyhedronAt50.contains(shouldBeOutside))
    }

    @Test
    fun testGetBoundingBox3D() {
        val boundingBox3D = tinyPolyhedronAtZero.boundingBox3D
        Assertions.assertTrue(boundingBox3D.contains(centerZero))
        Assertions.assertTrue(boundingBox3D.contains(pointInsideTinyPolyhedron))
        Assertions.assertFalse(boundingBox3D.contains(pointOutSideTinyPolyhedron))
    }
}
